### PR TITLE
Training and Inference on sub-domains

### PR DESCRIPTION
**Pull request by chelseajohn**
This PR contains code to perform sampling along with training and evaluation using patches(sub-domain) of samples.

## Training
**Training** on sub-domains is enabled by the arguments in `config.data`
```
data:
  use_domainSampling: true
  use_fixedPatchSize: false        # fixed domain with overlapping
  use_orderedSampling: false       # fixed domain w/o overlapping 
  pad_to_fullGrid: false
  use_fixedPatch_startIdx: true
  nPatch_per_sample: 100
  use_minLimit: true
  padding : [1,1,1,1]
  # slices: []
  # patch_startIdx: []
```

- With a combination of the above parameters the following sampling can be done:
  1. `nPatch_per_sample` random sized patches (sX,sY) randomly sampled from (nX,nY) grid (can contain overlapping patches)
      ```
      use_domainSampling: true 
      nPatch_per_sample: $value
      ```
      
  2. `nPatch_per_sample` fixed sized patch of (sX,sY) randomly sampled from (nX,nY) grid (can contain overlapping patches)
  
      ```
      use_domainSampling: true 
      nPatch_per_sample: $value
      use_fixedPatchSize: true
      ```
  3. To sample fixed sized patch of (sX,sY) without overlapping from (nX,nY) grid
      ```
      use_domainSampling: true 
      nPatch_per_sample: $value
      use_fixedPatchSize: true
      use_orderedSampling: true
      ```
  
      `nPatch_per_sample` will be overwritten to `(nX//sX)*(nY//sY)`. 
      ```
      slices: [[sX,sY]]
      ```

- Use `padding` argument the inflow/outflow of patch can be set.
- Use `use_minLimit: true` to restrict (sX,sY) to be > (2 * (kX-1), 2 * (kY-1)), it must be set to `True` when `pad_to_fullGrid` is `False`
- Use `use_fixedPatch_startIdx` to sample `slices` for every epoch from the same index
- Use `pad_to_fullGrid` to embed patches (sX,sY) into a (nX,nY) grid of zeroes 
- Additionally, to make training runs reproducible `slices` and `patch_startIdx` can be passed 

## Inference
To perform **evaluation** such that FNO predicts patches of the input grid , the model filters out the output patch before passing it into the `Q` projection layer. It can be enabled using the following arguments in `config.model`
```
model:
  get_subdomain_output: true
  iXBeg: 5
  iYBeg: 12
  iXEnd: 30
  iYEnd: 60

```
Keep in mind that the slice size `i{X,Y}End - i{X,Y}Beg` has to be > `2*(fourier_modes -1)`.

## Contour Plotting
`11_modelOutput.py` has been modified to allow evaluation and plotting contours for :
1. Full grid input (nx,ny) ---> full grid output (nx,ny) (`--use_full_input`)
2. Full grid input (nx,ny) ---> sub-domain grid output (sx,sy) (`--use_full_input --get_subdomain_output` + above mentioned `config.model` args)
3. patch grid input (sx,sy)---> patch grid output (sx,sy)  (`--iXBeg, --iXEnd, --iYBeg, --iYEnd`)

### CODE REVIEW

File: cfno/data/preprocessing.py [hunk 1]
```diff
@@ -110,18 +110,117 @@ def get_concat_data(self, task:str, nsamples:int, reader, multistep:bool=True):
...
+            sX, sY = self.slices[iPatch]
+            if len(self.slices) == 1:
+                xPatch_start = self.xPatch_start.copy()
+                yPatch_start = self.yPatch_start.copy()
+            else:
+                if self.use_ordered_sampling:
+                    xPatch_start = (iPatch // (self.nX//sX)) * sX
+                    yPatch_start = (iPatch % (self.nY//sY)) * sY            
+                else:
+                    xPatch_start = random.randint(0, self.nX - sX)
```
Comment (tlunet) [inline cfno/data/preprocessing.py]: It's very hard to understand what you are doing here ... you do pre-compute some the patche sizes, but not their positions ? So if I understood it correctly, calling  `__getitem__(idx)` twice using the same value for `idx` will not produce the same result if you use the random patch sampling ? 

If yes, that is a bit problematic, since we get a training dataset that is changing with the different epochs ...

File: cfno/data/preprocessing.py [hunk 2]
```diff
@@ -110,18 +110,117 @@ def get_concat_data(self, task:str, nsamples:int, reader, multistep:bool=True):
...
         return len(self.inputs)
 
     def __getitem__(self, idx):
-        inpt, outp = self.sample(idx)
+        if self.use_domain_sampling:
+            patch_padding = self.padding.copy()
+            iSample = idx // self.nPatch_per_sample
+            iPatch = idx % self.nPatch_per_sample
+            inpt_grid, outp_grid = self.sample(iSample)  
+            inpt, outp = np.zeros_like(inpt_grid), np.zeros_like(outp_grid)
```
Comment (tlunet) [inline cfno/data/preprocessing.py]: Don't understand why you need a copy here ... in my mind, you simply have to return only some patches from `inpt, outp` and not the same sample padded with zeros. This is very weird ...

File: cfno/models/cfno2d.py [hunk 3]
```diff
@@ -373,10 +386,16 @@ def forward(self, x):
                  x = self.channel_mlp[index](x) + x_skip_channel_mlp
                  if index < len(self.layers) - 1:
                     x = nn.functional.gelu(x)
+        
+        # to get only a subdomain output
+        if self.get_subdomain_output:
```
Comment (tlunet) [inline cfno/models/cfno2d.py:443]: Good idea to slice the space grid before upscaling !

File: scripts/11_modelOutput.py [hunk 4]
```diff
@@ -40,6 +55,21 @@
 refScales = args.refScales
 saveFig = args.saveFig
 
+# Need model config to get output of shape different to input
+# .i.e when using args.get_subdomain_ouput
```
Comment (tlunet) [inline scripts/11_modelOutput.py:59]: There is no need of setting the model differently if you want a subdomain output : use the same model whether you want it or not, and only restrict to grid subdomain at the end using `iXBeg`, etc ... 

Also, default value for `iXEnd` and `iYEnd` must be `None`, since `array[iBeg:]` is equivalent to `array[iBeg:None]`

File: scripts/11_modelOutput.py [hunk 5]
```diff
@@ -49,26 +79,70 @@
...
+    model = FourierNeuralOp(model=modelConfig, checkpoint=checkpoint)
+else:
+    model = FourierNeuralOp(checkpoint=checkpoint)
 
-model = FourierNeuralOp(checkpoint=checkpoint)
-uPred = model(u0)[varChoices.index(var)].T.copy()
+uPred = model(input)[varChoices.index(var)].copy()
+print(f'Shape of uInit: {uInit.shape}, uRef:{uRef.shape}, uPred: {uPred.shape}')
+
+xExpandedGrid = np.linspace(iXBeg, iXEnd, uPred.shape[0] + 1)
```
Comment (tlunet) [inline scripts/11_modelOutput.py]: Don't re-create the grid (especially since it is not uniform in Y direction for RBC). Use `xGrid, yGrid` defined above, and extract their subset using `iXBeg, ...`

File: scripts/11_modelOutput.py [hunk 6]
```diff
@@ -49,26 +79,70 @@
...
 if dataset.outType == "update":
     uRef /= dataset.outScaling
-if outType == "solution" and dataset.outType == "update":
-    uRef += uInit
-if outType == "update" and dataset.outType == "solution":
-    uRef -= uInit
 
-if outType == "update":
-    uPred -= uInit
+if uRef.shape != uInit.shape:
```
Comment (tlunet) [inline scripts/11_modelOutput.py:111]: This padding is very confusing ... why do you need that ?

File: scripts/configs/config_dte-3.yaml [hunk 7]
```diff
@@ -35,6 +39,11 @@ model:
   fno_skip_type: soft-gating
   use_postfnochannel_mlp: false
   channel_mlp_skip_type: 'soft-gating'
```
Comment (tlunet) [inline scripts/configs/config_dte-3.yaml:38]: Are `iXBeg`, ... used for the subdomain output, or for the input patch ? (or both ?) This is confusing ...

**Review by tlunet — CHANGES_REQUESTED** [review]
Hi @chelseajohn, thanks for the PR. There is many things that are not very clear, sometime confusing, that need to be properly defined. In particular, this parameters mixing between input patches and subdomain output. I believe this can be made much more simpler, which would avoid a lot of confusion later on.

File: cfno/data/preprocessing.py (hunk 1 above)
Comment (chelseajohn) [reply cfno/data/preprocessing.py]: Yes,that is true. Across epochs only the patch size is consistent and not the positions which will be random. On running with same patch size and position across epochs, the loss did not converge but the current strategy did.

File: cfno/data/preprocessing.py (hunk 2 above)
Comment (chelseajohn) [reply cfno/data/preprocessing.py]: PyTorch Dataloader does not allow for data to have varying sizes, so I emebdded the patches into full grid.

File: cfno/data/preprocessing.py (hunk 1 above)
Comment (tlunet) [reply cfno/data/preprocessing.py]: If the loss did not converge using fixed patch size and position, doesn't it mean that you need more patch per samples ?

Comment (chelseajohn) [reply cfno/data/preprocessing.py]: I tried only till 200 patches. I will try again with more and see if it works.

File: cfno/data/preprocessing.py (hunk 2 above)
Comment (tlunet) [reply cfno/data/preprocessing.py]: Oh, that's a critical point indeed ... because then the FNO doesn't have any information on the patch bounds when doing a forward pass on several batches. We need to find another solution, because now the FFT is applied on the whole subdomain with a lot of zeros, which should not be like this.

File: scripts/11_modelOutput.py (hunk 4 above)
Comment (chelseajohn) [reply scripts/11_modelOutput.py:59]: If you want the model to predict output of different size then you need to change the model last layer. If I do not change the model then the model is still predicting the input_size and we are just sampling using `iXBeg` and other paramters, so it does not make sense to say the model predicting subdomain given full domain.

It was easiest means to supply args `iBeg` and rest to model by using the model_config.

File: scripts/11_modelOutput.py (hunk 6 above)
Comment (chelseajohn) [reply scripts/11_modelOutput.py:111]: When you have input and output of diferent shape, you cannot perform  addition and subtraction.

File: scripts/configs/config_dte-3.yaml (hunk 7 above)
Comment (chelseajohn) [reply scripts/configs/config_dte-3.yaml:38]: It is used only when `--get_subdomain_output` is set to true for inference and it is applied to get a different ouput size than to input. It is not used on the input patch.

File: cfno/data/preprocessing.py (hunk 1 above)
Comment (tlunet) [reply cfno/data/preprocessing.py]: Well, it seems there is many possible ways of doing it. Ideally, it would be good to choose the patching setup using a named strategy, that can be described with one unique acronym, e.g :

- RS-RP : Random Size / Random Position (possibly overlapping, but fixed for each sample and not dynamic)
- RS-RPD : Random Size / Random Position Dynamic (dynamic, always)
- RS : Random Split (non overlapping split of the whole domain)
- etc ... 

Then a small utility method read the string, and determine the whole strategy from it. First splitting with "-" and seeing if there is two terms (size/ position are computed randomly) or one term (domain is split randomly, and size / position are determined from it), then looking if there is a D at the end of the term or not for dynamic, etc ... And all the magic should be done internally by the Dataloader.

But we still have to find a proper solution for the batch with variable grid size ...

File: cfno/data/preprocessing.py (hunk 2 above)
Comment (tlunet) [reply cfno/data/preprocessing.py]: I may have a solution ... will detail that on Mattermost

File: scripts/11_modelOutput.py (hunk 5 above)
Comment (chelseajohn) [reply scripts/11_modelOutput.py]: Done!

File: scripts/11_modelOutput.py (hunk 6 above)
Comment (tlunet) [reply scripts/11_modelOutput.py:111]: Well, theoretically that's how I would see that (on a 1D domain to simplify ...)

```python
# Retrieve fields on the whole subdomain
inpt, outp = dataset.sample(idx)

# Restrict to patch
inpt = inpt[:, iBeg:iEnd]
outp = outp[:, iBeg:iEnd]

# Evaluate model
uPred = model(inpt)

# And do the addition / subtraction to adapt depending on whether we want the solution or the update ...
```

So I don't understand why you need this padding. This is probably related to the topic I've just written on Mattermost, we may need to discuss that per zoom ...

Comment (chelseajohn) [reply scripts/11_modelOutput.py:111]: For the correction of update and solution, I realised I could have made uInit into uRef shape which is easier than doing the other way around. I have pushed the change.
